---
title: "revision"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##R Markdown

```{r}
library(Seurat)
library(ggplot2)
```

Load in the MG dataset: 
```{r}
load("/Users/whippoorwill/Desktop/Sequencing/LD_AVM02/Data/Seurat/mgAVM02_donotdelete.RData")
```
```{r}
DimPlot(mgAVM02,group.by = "seurat_clusters")
DimPlot(mgAVM02,group.by = "finalclusters")
```

Question 1: Where are clusters 5 and 7? Tree is saved as /Users/whippoorwill/Dropbox (Anna Molofsky Lab)/2020-Leah-barrelCortex/Manuscript data/Glia RNASeq/P5_P7 MG rnaseq/trees/tree_MGonlyMarch2020WITH57.eps

merged 5 with 0, 6 with 7
```{r}
library(ape)
PlotClusterTree(mgAVM02, cex=1, use.edge.length=T,node.depth=2)

tree = mgAVM02@tools$BuildClusterTree
setEPS()
postscript("~/Desktop/tree_mgAVM02.with57.eps")
plot.phylo(tree, use.edge.length = T, direction = "rightwards")
dev.off()

```


Load in spreadsheet containing original cluster defining genes: 
```{r}
markers = read.csv("/Users/whippoorwill/Desktop/Sequencing/LD_AVM02/P5-P7 from dropbox/New plots march 24 2020/allmarkers_vargenesMG_0324.csv")

markers = read.csv("/Users/whippoorwill/Desktop/Sequencing/LD_AVM02/Spreadsheets/MAST_newthresholds_Feb2020/allmarkers_vargenesMG_0215.csv")
```

Number of cells per cluster and DE genes per cluster
```{r}
print("original")
table(mgAVM02$seurat_clusters)
print("modified")
table(mgAVM02$finalclusters)

markers = markers[markers$p_val_adj < 1e-5,]
markers = markers[markers$avg_logFC > 0.2,]
print("Marker genes per cluster")
table(markers$cluster)
```
Note that by these standards (all markers) 5 and 7 look fine. But if you look for single markers: 

```{r}
singlemarkers = markers[markers$gene %in% names(table(markers$gene))[table(markers$gene) ==1],]
```
Print out single markers (unique cluster-defining markers)

```{r}
table(singlemarkers$cluster)
write.csv(table(singlemarkers$cluster),file = "~/Desktop/singlemarkersbeforecombination.csv")
```
Note that clusters 0/5, 7/6 have few (<50) unique markers
Returning to the tree, note that 5+0 are closely related; 6+7 are closely related. So I started by combining 5,0 and 6,7 as they didn't have enough unique markers to call them separate clusters. 



Load in markers with 0/5 and 6/7 combined: 
```{r}
markers = read.csv("/Users/whippoorwill/Desktop/Sequencing/LD_AVM02/Spreadsheets/allmarkers_vargenesMG_0501 copy.csv")

markers = markers[markers$p_val_adj < 1e-3,]
markers = markers[markers$avg_logFC > 0.25,]
singlemarkers = markers[markers$gene %in% names(table(markers$gene))[table(markers$gene) ==1],]
print("Marker genes per cluster")
table(singlemarkers$cluster)
write.csv(table(singlemarkers$cluster),file = "~/Desktop/singlemarkerspostcombination.csv")

```
Note that 5/6 still have few markers, but they are not **closely related** to any clusters with few markers. 

2. Gunner et al dataset
  Note - P5, papain, 823 cells, no interferon cluster, lots of the "neuronal engulfing" cluster (15-20%) in both control and deprived. 
  
see singlecellmodels code sheet. 

3. % of cells with neuronal transcripts

Neuronal transcripts: 

```{r}
markers = read.csv("/Users/whippoorwill/Dropbox (Anna Molofsky Lab)/2020-Leah-barrelCortex/DATA/Glia RNASeq/P5_P7 MG rnaseq/Spreadsheets/allmarkers_vargenesMG_0501.csv")

markers = markers[markers$cluster == '3',]
markers = markers[markers$p_val_adj<1e-10,]
markers = markers[markers$avg_logFC > 0.25,]

c3genes = markers$gene

```

Compare with Barres dataset and show enrichment in oligos/astrocytes/microglia/neurons
```{r}
barres = read.csv("/Users/whippoorwill/Desktop/Code for paper/Necessary spreadsheets/barreslab_rnaseq.csv")
barres = barres[barres$Gene.symbol %in% c3genes,]
dim(barres) #207 out of 238 genes
length(c3genes)
head(barres)
```
Find out which gene is specific to which cell type by making a new column called "enrichment" that names the celltype with the most expression

```{r}
celltype = c(rep(0,nrow(barres)))

enrichment = c(rep(0,nrow(barres)))
for (i in 1:nrow(barres)){
  g = barres[i,3:9]
  m = max(g)
  c = names(g)[g == max(g)][1] #arbitrarily picking the first instance in case of a tie
  nextmax = max(g[names(g)!=c]) #next highest enrichment
  e = (m/nextmax) #overenrichment compared to next highest cell type
  celltype[i]= c
  enrichment[i]= e
}

barres = cbind(barres,celltype,enrichment)
```

Make a table showing how many are enriched from each cell type: 
```{r}
table(barres$celltype)

barres$celltype[barres$celltype %in% c("Newly.Formed.Oligodendrocyte","Myelinating.Oligodendrocytes","Oligodendrocyte.Precursor.Cell")] = "Oligodendrocytes"
g <- ggplot(barres, aes(celltype, enrichment))
g + geom_boxplot(varwidth=T, fill="plum") + 
    labs(title="Box plot", 
         subtitle="Enrichment vs Celltype",
         caption="Source: barres",
         x="Celltype",
         y="Enrichment")

setEPS()
postscript("~/Desktop/neuronalenrichment.eps")
g + geom_boxplot(varwidth=T, fill="plum") + 
    labs(title="Box plot", 
         subtitle="Enrichment vs Celltype",
         caption="Source: barres",
         x="Celltype",
         y="Enrichment")
dev.off()

```
Exclusive to genes >3x enriched
```{r}
barres = barres[barres$enrichment >3,]
table(barres$celltype)
```


Make a module score for each celltype
```{r}
anno = barres[,c(1,10)]
anno$celltype = as.factor(anno$celltype)
genelist = list()
for (cell in levels(anno$celltype)){
  genelist[[cell]] = anno$Gene.symbol[anno$celltype == cell]
}
genelist
```
Add module scores

```{r}
mgAVM02 = AddModuleScore(mgAVM02,features = genelist,name = names(genelist))

```
Plot each module
```{r}
i = 1
for (celltype in names(genelist)){
  n = paste0(celltype,i)
  print(FeaturePlot(mgAVM02,features = n,cells = colnames(mgAVM02)[mgAVM02$finalclusters == '3']))
  i = i+1
}
```
average neuronal module score in all cells: <0
```{r}
median(mgAVM02$Neuron3)
Idents(mgAVM02) = "finalclusters"
VlnPlot(mgAVM02,"Neuron3",log = F,pt.size = 0,idents = c("5","1","2","4","6","8"))
```
% of cells in cluster 3 with Neuron module score greater than 0.25
```{r}
cluster3modulescore = mgAVM02$Neuron3[mgAVM02$finalclusters == "3"]
length(cluster3modulescore)
length(cluster3modulescore[cluster3modulescore >0.25])

100*length(cluster3modulescore[cluster3modulescore >0.25])/length(cluster3modulescore)
#46% of cluster3 cells have a module score for the 45 neuronal genes > 0.25
```
In more concrete gene-specific terms: 

```{r}

#isolate the normalized counts for just the neuronal genes across all cells and find the mean, median: 
neuronalgenes = GetAssayData(mgAVM02,slot = "data",assay = "RNA")[genelist$Neuron,]
totalneuronalgenes = colSums(neuronalgenes)
mgAVM02$NeuronalGenesSum = totalneuronalgenes
setEPS()
postscript("~/Desktop/NeuronalGenesSumVln.eps")
VlnPlot(mgAVM02,"NeuronalGenesSum",group.by = "cl38only",pt.size = 0)
dev.off()
stats = do.call("cbind", tapply(mgAVM02$NeuronalGenesSum,mgAVM02$cl38only,quantile,probs=c(0.05,0.5,0.95)))
write.csv(stats,file = "~/Desktop/NeuronGeneSumStatsByCluster.csv")

anno_vln = function(seurat_object = sobject, metric = "nCount_RNA", group = "orig.ident",keep = "all" ){
  #specify number of categories
  Idents(seurat_object) = group
  n = length(levels(seurat_object@meta.data[,group]))
  if (keep!= "all"){n = length(keep)}

  #make a violin plot
  a = VlnPlot(seurat_object,metric,group.by = group,pt.size = 0,log = F,idents = keep)

  #make a simple function that quantifies percentiles for each column
  add.percentiles = function(x) {
    r = quantile(x, probs = c(0.05,0.5,0.95))
    r
  }

  a + geom_boxplot(width = 0.05,outlier.size = 0,notchwidth = 10,notch = T) + theme_bw() + stat_summary(fun=add.percentiles, geom="point", pch="_", colour=c(rep(c("forestgreen","darkblue","grey"),n)), size=10)
}

setEPS()
postscript("~/Desktop/NeuronalGenesSum_without39.eps")
anno_vln(mgAVM02,metric = "NeuronalGenesSum",group = "finalclusters",keep = c("1","2","4","5","6","8"))
dev.off()
write.csv(genelist$Neuron,file = "~/Desktop/neuronalgenesexclude.csv")


```


```{r}
VlnPlot(mgAVM02,"NeuronalGenesSum",group.by = "finalclusters")
max(mgAVM02$NeuronalGenesSum[!mgAVM02$finalclusters %in% c("3","8","9")]) #20.14
#percent of cells in cluster 3 with higher expression than *all other cells*: 
x = mgAVM02$NeuronalGenesSum[mgAVM02$finalclusters == "3"]
length(x) #1235
length(x[x>20.14]) #120
519/1235 #9.7% of cells in cluster 3 have more neuronal RNA than *all other cells*

quantile(mgAVM02$NeuronalGenesSum[!mgAVM02$finalclusters %in% c("3","8","9")],probs = 0.99) #11.77
length(x[x>11.77]) #1030
1030/1235 #83.4 % of cells in cluster 3 have more neuronal RNA than *99% of other cells*

quantile(mgAVM02$NeuronalGenesSum[!mgAVM02$finalclusters %in% c("3","8","9")],probs = 0.95) #8.53
length(x[x>8.53]) #1139
1142/1235 #92.5% % of cells in cluster 3 have more neuronal RNA than *90% of other cells*

y = mgAVM02$NeuronalGenesSum[mgAVM02$finalclusters == "8"]
length(y) #329
length(y[y>8.53]) #32
length(y[y>11.77]) #15
32/329 #9.7% of cluster 8 has more neuronal mrna than 95% of remaining cells
15/329 #4.6% of cluster 8 has more neuronal mRNA than 99% of remaining cells

```

```{r}
FeaturePlot(mgAVM02,features = c("Astrocytes1","Neuron5"),blend = T,order = T,cells = colnames(mgAVM02)[mgAVM02$finalclusters == '3'])

VlnPlot(mgAVM02,features = c("Astrocytes1","Neuron5"),group.by = "finalclusters")

d = mgAVM02@meta.data[colnames(mgAVM02)[mgAVM02$finalclusters == '3'],c("Astrocytes1","Endothelial.Cells2","Myelinating.Oligodendrocytes4","Neuron5","Newly.Formed.Oligodendrocyte6","Oligodendrocyte.Precursor.Cell7")]

plot(x = d$Astrocytes1,y = d$Neuron5)
plot(x = d$Myelinating.Oligodendrocytes4,y = d$Neuron5)
plot(x = d$Newly.Formed.Oligodendrocyte6,y = d$Neuron5)
plot(x = d$Oligodendrocyte.Precursor.Cell7,y = d$Neuron5)
plot(x = d$Endothelial.Cells2,y = d$Neuron5)
plot(x = d$Astrocytes1,y = d$Endothelial.Cells2)
```
```{r}
g <- ggplot(barres, aes(celltype, enrichment))
g + geom_boxplot(varwidth=T, fill="plum") + 
    labs(title="Box plot", 
         subtitle="Enrichment vs Celltype",
         caption="Source: barres",
         x="Celltype",
         y="Enrichment")

```
Notes from this analysis: *ALL* the uniquely enriched (<5x) genes are neuronal; many are shared between multiple celltypes but enriched compared to microglia, so clearly foreign but also not necessarily helpful in determining celltype specificity. All signs point to neuronal engulfment. 



5. Pseudotime (specifically looking for any connection between clusters 3 and 8) - nothing
```{r}
library(monocle3)
```

```{r}
expression_matrix <- GetAssayData(mgAVM02,assay = "RNA",slot = "counts")
cell_metadata <- mgAVM02@meta.data
gene_annotation <- cbind(rownames(mgAVM02),"gene_short_name" = rownames(mgAVM02))
rownames(gene_annotation) = rownames(mgAVM02)

cds <- new_cell_data_set(expression_matrix,
                         cell_metadata = cell_metadata,
                         gene_metadata = gene_annotation)
```

```{r}
cds <- preprocess_cds(cds, num_dim = 50,)
cds <- align_cds(cds, residual_model_formula_str = "~ percent.mito + nCount_RNA")
```
```{r}
cds <- reduce_dimension(cds,reduction_method = "UMAP")
```


```{r}
setEPS()
postscript("~/Desktop/mgavm02_clusters_monocle.eps")
plot_cells(cds, label_groups_by_cluster=T,  color_cells_by = "finalclusters",group_label_size = 5,show_trajectory_graph = F)
dev.off()

setEPS()
postscript("~/Desktop/mgavm02_age_monocle.eps")
plot_cells(cds, label_groups_by_cluster=T,  color_cells_by = "age",group_label_size = 5)
dev.off()
```

```{r}
genes <- c("Ifitm3","Ifit3","P2ry12","Nrxn3","Meg3")

plot_cells(cds,
           genes=genes,
           label_cell_groups=FALSE,
           show_trajectory_graph=FALSE)
```

```{r}
cds <- cluster_cells(cds,cluster_method = "leiden")
plot_cells(cds, color_cells_by = "partition")
```

```{r}
cds <- learn_graph(cds)
plot_cells(cds,
           color_cells_by = "finalclusters",
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE)
```

```{r}
cds <- order_cells(cds)
```
```{r}
setEPS()
postscript("~/Desktop/pseudo_mgAVM02_all.eps")
plot_cells(cds,
           color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=4)
dev.off()
```
```{r}
# a helper function to identify the root principal points:
get_earliest_principal_node <- function(cds, time_bin="Juvenile"){
  cell_ids <- which(colData(cds)[, "age"] == time_bin)
  
  closest_vertex <-
  cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])
  root_pr_nodes <-
  igraph::V(principal_graph(cds)[["UMAP"]])$name[as.numeric(names
  (which.max(table(closest_vertex[cell_ids,]))))]
  
  root_pr_nodes
}
cds <- order_cells(cds, root_pr_nodes=get_earliest_principal_node(cds))
```
```{r}
plot_cells(cds,
           color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=4)
```

```{r}
library(sf)
plot_cells(cds,
           color_cells_by = "namedclusters",
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE)

test_res <- graph_test(cds, neighbor_graph="principal_graph", cores=4)
```


```{r}
pr_deg_ids <- row.names(subset(test_res, q_value < 0.005))
head(pr_deg_ids)

plot_cells(cds, genes=head(pr_deg_ids),
           show_trajectory_graph=FALSE,
           label_cell_groups=FALSE,
           label_leaves=FALSE)
```
```{r}
VlnPlot(mgAVM02,c("Map1b","Arpp21","Ahi1","Ptprd","Pfkp","Nrxn1","Gria2","Ank3"),group.by = "finalclusters")
```
Compare cluster 3 + 8 vs all other

```{r}
markers38vall = FindMarkers(mgAVM02,ident.1 = c("3","8"),group.by = "finalclusters",test.use = "MAST",features = rownames(mgAVM02),assay = "RNA",slot = "counts",logfc.threshold = 0.2,min.pct = 0.2)
```
```{r}
markers38vall = markers38vall[order(markers38vall$avg_log2FC,decreasing = T),]

```
Load in markers 

```{r}
markers = read.csv("/Users/whippoorwill/Dropbox (Anna Molofsky Lab)/2020-Leah-barrelCortex/Manuscript data/Glia RNASeq/P5_P7 MG rnaseq/Spreadsheets/allmarkers_vargenesMG_0501.csv")
markers = markers[markers$cluster %in% c('3','8'),]

markers = markers[markers$p_val_adj<1e-3,]
markers = markers[markers$avg_logFC>0.2,]

markers38vall = markers38vall[markers38vall$avg_log2FC>0.25,]
markers38vall = markers38vall[markers38vall$p_val_adj<1e-5,]



```

```{r}
markers
markers38vall

rownames(markers38vall[!rownames(markers38vall)%in% markers$gene,])
markers38vallexcl = markers38vall[!rownames(markers38vall) %in% markers$gene,]
markers$gene

DoHeatmap(
    object = mgAVM02, 
    features = rownames(markers38vallexcl)[1:100],
    group.by = 'finalclusters',
    size = 5,
    label = T,
    draw.lines = T
)
```
```{r}
markers = read.csv("/Users/whippoorwill/Dropbox (Anna Molofsky Lab)/2020-Leah-barrelCortex/Manuscript data/Glia RNASeq/P5_P7 MG rnaseq/Spreadsheets/allmarkers_vargenesMG_0501.csv")
markers = markers[markers$cluster %in% c('8','3'),]
markers = markers[markers$avg_logFC>0.1,]
markers = markers[markers$p_val_adj<1e-3,]
table(markers$cluster)

sharedgenes = markers$gene
sharedgenes = sharedgenes[order(sharedgenes)]
both = sharedgenes[duplicated(sharedgenes)]
both
```

```{r}
eightgenes = read.csv("/Users/whippoorwill/Desktop/Jupyter/cluster8genesvelocity.csv")
eightgenes = eightgenes$X8
```


```{r}
for (gene in eightgenes[1:50]){
  print(FeaturePlot(mgAVM02,gene))
}

```
Male vs female:

```{r}
#Female: Xist, Tsix, 
#Male: Eif2s3y, Ddx3y
print(plot(x = GetAssayData(mgAVM02,slot = "counts")["Xist",],y=GetAssayData(mgAVM02,slot = "counts")["Eif2s3y",]))
print(VlnPlot(mgAVM02,c("Xist","Tsix","Eif2s3y","Ddx3y"),group.by = "sex"))
```

#Assign cell names of "Male" or "Female" depending on Xist and Ddx3y/Eif2s3y expression
```{r}
#identify cell names with Xist expression
mgAVM02$malegenes = colSums(GetAssayData(mgAVM02,slot = "counts")[c("Ddx3y","Eif2s3y"),])
mgAVM02$femalegenes = colSums(GetAssayData(mgAVM02,slot = "counts")[c("Xist","Tsix"),])

female<-WhichCells(object = mgAVM02, cells = colnames(mgAVM02)[mgAVM02$femalegenes > 0 & mgAVM02$malegenes ==0],slot = "counts") 

male<-WhichCells(object = mgAVM02, cells = colnames(mgAVM02)[mgAVM02$femalegenes == 0 & mgAVM02$malegenes > 0],slot = "counts") 
length(female)
length(male)

mixed = WhichCells(object = mgAVM02, cells = colnames(mgAVM02)[mgAVM02$femalegenes > 0 & mgAVM02$malegenes > 0],slot = "counts") 
neither = WhichCells(object = mgAVM02, cells = colnames(mgAVM02)[mgAVM02$femalegenes == 0 & mgAVM02$malegenes == 0],slot = "counts") 

#set up a vector with the correct number of "male" and "female" labels
sex=as.factor(c(rep("Female",length(female)),rep("Male",length(male)),rep("mixed",length(mixed)),rep("neither",length(neither)))) 

#assign each cell ID to its appropriate label; must match order set up in the vector above.
names(sex)<-c(female,male,mixed,neither) 
table(sex)

#order this vector by the cellnames vector found in names(DG@meta.data)
cellnames = rownames(mgAVM02@meta.data)
sex = sex[match(cellnames,names(sex))]
sex = as.character(sex)
sex[is.na(sex)] = "Unknown"

table(sex)
#Add in a metadata column for the sex of the cell/animal
mgAVM02$sexstringent <- sex
mgAVM02$sexstringentcondition = paste0(mgAVM02$sexstringent,mgAVM02$sample_description)
#Check that the numbers match with the table above
table(mgAVM02$sexstringent) 
table(mgAVM02$sexstringent,mgAVM02$age)
x = table(mgAVM02$sexstringentcondition,mgAVM02$finalclusters)
x = cbind(x,"total" = rowSums(x))
x = cbind(x,"cluster8pct" = 100*x[,7]/x[,9])
x = cbind(x,"cluster3pct" = 100*x[,4]/x[,9])
x

write.csv(x,file = "~/Desktop/clusterpctstringentsex.csv")

DimPlot(mgAVM02,split.by = "sample_description",group.by = "sexstringent")
DimPlot(mgAVM02,group.by = "sample_description",split.by = "sexstringent")
cells = as.factor(mgAVM02$sexstringent)
keep = c("Male","Female")
cells = cells[cells %in% keep]
cells = droplevels(cells)
newcells = c()
for (level in levels(cells)){
  n = sample(names(cells)[cells == level],1500)
  newcells = c(newcells,n)
}

setEPS()
postscript("~/Desktop/stringentsexDimPlots.eps")
DimPlot(subset(mgAVM02,cells = newcells),group.by = "sample_description",split.by = "sexstringent")
dev.off()

setEPS()
postscript("~/Desktop/stringentsexDimPlots2.eps")
DimPlot(subset(mgAVM02,cells = newcells),group.by = "sexstringent")
dev.off()

setEPS()
postscript("~/Desktop/stringentsexvlnplot.eps")
VlnPlot(mgAVM02,c("Xist","Tsix","Eif2s3y","Ddx3y"),group.by = "sexstringent",pt.size = 0)
dev.off()
```
```{r}
#Doublet Finder on total dataset
library(modes)
library(DoubletFinder)

for (sample in levels(as.factor(mgAVM02$sample_description))){
  LD1Ctrl = subset(mgAVM02,subset = sample_description == "Control_P5")
  sweep.res.listC <- paramSweep_v3(LD1Ctrl, PCs = 1:30, sct = FALSE)
  sweep.statsC <- summarizeSweep(sweep.res.listC, GT = FALSE)
  bcmvnC <- find.pK(sweep.statsC)
  homotypic.prop <- modelHomotypic(mgAVM02@meta.data$finalclusters) 
  nExp_poi <- round(0.075*nrow(mgAVM02@meta.data))  ## Assuming 7.5% doublet formation rate - tailor for your dataset
  nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
  mgAVM02 = doubletFinder_v3(mgAVM02,PCs = 1:30, pN = 0.25, pK = 0.09, nExp = nExp_poi.adj,reuse.pANN = F, sct = F)
  
}



setwd("~/Desktop")



table(mgAVM02$DF.classifications_0.25_0.09_925)
x = table(mgAVM02$DF.classifications_0.25_0.09_925,mgAVM02$sex)
x = rbind(x,"total" = colSums(x))
x = rbind(x,"pctdoublet"= 100*x[1,]/x[3,])
x

mgAVM02$sexdoublet = paste0(mgAVM02$DF.classifications_0.25_0.09_925,"_",mgAVM02$sex)
VlnPlot(mgAVM02,c("Xist","Ddx3y"),group.by = "sexdoublet")

mgAVM02$sexdoubletcondition = paste0(mgAVM02$sexdoublet,mgAVM02$sample_description)
x = table(mgAVM02$sexdoubletcondition,mgAVM02$finalclusters)
x = cbind(x,"total" = rowSums(x))
x[c(13,15,17,19),]
x[c(14,16,18,20),]

x = table(mgAVM02$sex,mgAVM02$finalclusters)
x = rbind(x,"total" = colSums(x))
x = rbind(x,"pctdoublet" = 100*x[3,]/x[4,])
x
```


```{r}
#Create a new metadata column combining sample_description [or in your case Il33 genotype, whatever that is named in your metadata] and sex: 
fulldescription = paste0(mgAVM02$sample_description, "_",mgAVM02$sex)
names(fulldescription) = names(mgAVM02@meta.data)
table(fulldescription)
mgAVM02$fulldescription = fulldescription

table(mgAVM02$fulldescription,mgAVM02$finalclusters)
table(mgAVM02$sex,mgAVM02$sample_description)
```
```{r}
setEPS()
postscript("~/Desktop/vlnMF.eps")
VlnPlot(mgAVM02,c("Xist","Tsix","Eif2s3y","Ddx3y"),group.by = "sex",pt.size = 0)
dev.off()


setEPS()
postscript("~/Desktop/umapMF.eps")
DimPlot(mgAVM02,split.by = "sample_description",group.by = "sex")
dev.off()

cellnames = as.factor(mgAVM02$fulldescription)
c = NULL
for (s in levels(cellnames)){
  cells = cellnames[cellnames == s]
  cells = sample(names(cells),500)
  c = c(c,cells)
}
length(c)
setEPS()
postscript("~/Desktop/umapMF.eps")
DimPlot(mgAVM02,split.by = "sample_description",group.by = "sex")
dev.off()

setEPS()
postscript("~/Desktop/umapMF-equal.eps")
DimPlot(mgAVM02,cells = c,group.by = "sex")
dev.off()

save(mgAVM02,file = "~/Desktop/Sequencing/LD_AVM02/Data/Seurat/mgAVM02_donotdelete.RData")
```
4. Analyze Syage, Frigerio datasets, + any models discussed that have scRNAseq datasets (lol I took those out before...)
ProjectTILs to overlay my single cell with other people's data

```{r}
load("/Users/whippoorwill/Desktop/Sequencing/LD_AVM02/Data/Seurat/mgAVM02_donotdelete.RData")
```
```{r}
library(Seurat)
#BiocManager::install("umap")
library(umap)
library(ggplot2)
```

Re-using existing pc's; not convinced this actually works

```{r}
ref.pca = mgAVM02@reductions$pca@cell.embeddings
ref.pca[1:5,1:5]
```
Subset the object dramatically
```{r}
cellnames = sample(colnames(mgAVM02),4000)
ref= subset(mgAVM02,cells = cellnames)
DimPlot(ref)
```

```{r}
seed=42
n.neighbors=30
min.dist=0.3
metric="cosine"
ndim=30
```

```{r}
which.assay="RNA"
varfeat <- ref@assays[[which.assay]]@var.features
  

refdata <- data.frame(t(as.matrix(ref@assays[[which.assay]]@data[varfeat,])))

refdata <- refdata[, sort(colnames(refdata))]
refdata = refdata[,colSums(refdata) >10]
min(colSums(refdata)) #greater than 0

ref.pca <- prcomp(refdata, rank. = 50, scale. = TRUE, center = TRUE, retx=TRUE)

ref.pca$rotation[1:5,1:5]
```


```{r}
umap.config <- umap.defaults
umap.config$n_neighbors = n.neighbors
umap.config$min_dist = min.dist
umap.config$metric = metric
umap.config$n_components = 2
umap.config$random_state = 4
umap.config$transform_state = 100

ref.umap <- umap(ref.pca$x[,1:ndim], config=umap.config)

colnames(ref.umap$layout) <- c("UMAP_2","UMAP_1")
ref.umap
```


```{r}
ref@reductions$umap@cell.embeddings <- ref.umap$layout

#Store the complete PCA and UMAP object in @misc
ref@misc$pca_object <- ref.pca
ref@misc$umap_object <- ref.umap
ref@misc$projecTILs="custom_atlas"
```
```{r}
DimPlot(ref, reduction = "umap", pt.size = 0.5, group.by = "finalclusters") + ggtitle("UMAP by cluster")

```


```{r}
saveRDS(ref,file = "~/Desktop/Sequencing/LD_AVM02/Data/Seurat/mgAVM02_customUMAP_newPCA.rds")
```

if (!requireNamespace("remotes")) install.packages("remotes")
library(remotes)

if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if (!requireNamespace("Seurat", quietly = TRUE)) {
   install.packages("Seurat")
}

if (!requireNamespace("TILPRED", quietly = TRUE)) {
  install.packages(c("doParallel","doRNG"))
  BiocManager::install(c("AUCell","SingleCellExperiment"))
  remotes::install_github("carmonalab/TILPRED")
}

remotes::install_github("carmonalab/ProjecTILs")


**Start here***
```{r}
library(ProjecTILs)
ref <- load.reference.map(ref="~/Desktop/Sequencing/LD_AVM02/Data/Seurat/mgAVM02_customUMAP_newPCA.rds")

ref@assays$integrated = ref@assays$RNA
```

Go to https://carmonalab.github.io/ProjecTILs/tutorial.html for instructions on embedding new data into this umap

#A sample data set is provided with the ProjecTILs package


```{r}
#querydata <- ProjecTILs::query_example_seurat
```


More generally, it is possible to load a query matrix with gene names and barcodes (e.g. 10X format or raw counts)
```{r}
#1. Frigerio
##10X format
filename <- "~/Desktop/revisions/sequencing/single cell models/frigerio/raw/GSE127884_microglia.cdep.raw.tsv.gz"
filename2 <- "~/Desktop/revisions/sequencing/single cell models/frigerio/raw/GSE127892_microglia.kw.raw.tsv.gz"
#filename = "/Users/whippoorwill/Desktop/single cell models/5xFAD/GSE148405_counts.csv.gz"
querydata <- read.sc.query(filename, type = "raw")
querydata2 <- read.sc.query(filename2, type = "raw")

#for a list of folders
filename = "/Users/whippoorwill/Desktop/single cell models/GBM/GSE136001_RAW/"
folders = list.files(path = filename)
folders
l = list()
for (folder in folders){
  path = paste0(filename,folder)
  querydata = read.sc.query(path,type = "10x")
  querydata@meta.data$orig.ident = folder
  l[[folder]] = querydata
}
```


```{r}
#for a list of folders
filename = "/Users/whippoorwill/Desktop/revisions/sequencing/single cell models/tMCAO/GSE171169_RAW/"
folders = list.files(path = filename)
folders
l = list()
for (folder in folders){
  path = paste0(filename,folder)
  querydata = read.sc.query(path,type = "10x")
  querydata@meta.data$orig.ident = folder
  l[[folder]] = querydata
}

```

Combine the datasets
```{r}
sobject = l[[1]]
for (i in 2:length(l)){
  sobject = merge(sobject,l[[i]])
}


#Filter:

sobject = PercentageFeatureSet(sobject,pattern = "^mt-",col.name = "pct.mito")
VlnPlot(sobject,c("nCount_RNA","pct.mito"),group.by = "orig.ident") + geom_abline(intercept = 5, slope = 0)

sobject = subset(sobject,subset = pct.mito<5)
VlnPlot(sobject,c("nFeature_RNA","pct.mito"),group.by = "orig.ident",pt.size = 0) + geom_abline(intercept = 5, slope = 0)
sobject = subset(sobject,subset = nFeature_RNA<2000)
sobject = subset(sobject,subset = nFeature_RNA>500)

save(sobject,file = "~/Desktop/single cell models/GBM/GBMfiltered.RData")
save(sobject,file = "~/Desktop/single cell models/GBM/GBMfilteredsubset.RData")
```


Look at the data
```{r}
querydata
table(querydata$orig.ident)

querydata = subset(querydata,subset = orig.ident %in% c("P209", "P211"))

x = as.character(querydata$orig.ident)
x[x == "P209"] = "WT"
x[x == "P211"] = "5xFAD"
names(x) = colnames(querydata)
querydata$genotype = x
```

```{r}
head(colnames(querydata))
m = querydata@meta.data

metadata = read.csv("/Users/whippoorwill/Desktop/revisions/sequencing/single cell models/frigerio/metadata/GSE127884_microglia.cdep.meta.csv.gz",row.names = 1,stringsAsFactors = F)
head(rownames(metadata))

f2 = function(x){x = sub(' ','',x)}
rownames(metadata) = sapply(rownames(metadata),f2)
head(rownames(m))
head(rownames(metadata))
```


```{r}
length(rownames(m))
length(rownames(metadata))

metadata = metadata[match(rownames(m),rownames(metadata)),]

head(metadata)
head(m)
```
```{r}
identical(rownames(m),rownames(metadata)) #must be true

```
Add in metadata

```{r}
querydata@meta.data = metadata

head(querydata@meta.data)
head(m)
querydata@meta.data$nCount_RNA = m$nCount_RNA
querydata@meta.data$nFeature_RNA = m$nFeature_RNA
```
```{r}
querydata2 = subset(querydata2,cells = sample(colnames(querydata2),5000))
sobject = merge(querydata,querydata2)
```
```{r}
save(sobject,file = "~/Desktop/revisions/sequencing/single cell models/frigerio/combinedseurat.Rdata")
```

Run Projection algorithm

```{r}
cellnames = sample(colnames(sobject),10000)
sobject = subset(sobject,cells = cellnames)
```
Syage et al

```{r}
querydata = readRDS("/Users/whippoorwill/Desktop/AScombined_v10.7.rds")
querydata = subset(querydata, subset = celltype %in% c("Mac1","Mac2","Mac3","Mac4","MG1","MG2","MG3","Cyc. MG"))
DimPlot(querydata)
```


```{r}
query.projected <- make.projection(querydata, ref=ref,filter.cells = F,skip.normalize = F,human.ortho = F)
```



Plot projection of new data over the reference in UMAP space. The contour lines display the density of projected query cells onto the reference map.

```{r}
querycontrol = subset(query.projected,subset = orig.ident == "Control")
queryDay37 = subset(query.projected,subset = orig.ident %in% c("Day3","Day7"))
plot.projection(ref, query.projected,labels.col = "finalclusters") + ggtitle("all")
plot.projection(ref, querycontrol,labels.col = "finalclusters") + ggtitle("Control")
plot.projection(ref, queryDay37,labels.col = "finalclusters") + ggtitle("Day3 and Day 7 post infection")
```

```{r}
table(query.projected$orig.ident)


```
Label the cells that fall into my cluster 8 region: 

```{r}
umap = query.projected@reductions$umap@cell.embeddings
cells = rownames(umap[umap[,1]>3 & umap[,2]<0,])

allcells = colnames(query.projected)
allcells = rep("other",length(allcells))
head(allcells)
names(allcells) = colnames(query.projected)
allcells[names(allcells) %in% cells] = "cluster8"
query.projected$annotated = allcells


x = table(query.projected$annotated,query.projected$orig.ident)
rbind(x,"percent in interferon cluster" = 100*(x[1,]/(x[2,]+ x[1,])))

save(query.projected,file = "~/Desktop/COVIDprojected.RData")

```

So just repeat this for all the datasets
a) Syage et al (done) - coronavirus in mouse (not covid)
b) LPS Stevens (done) GSE121654 
d) DONE GBM https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE136001
h) Appswe.ps1de9 (done, frigerio)
i) 5xFAD (https://www-ncbi-nlm-nih-gov.ezproxy.u-pec.fr/geo/query/acc.cgi?acc=GSE161224
  also GSE 148405 https://www-ncbi-nlm-nih-gov.ezproxy.u-pec.fr/geo/query/acc.cgi?acc=GSE148405) - done
e) tMCAO GSE171169 https://www-ncbi-nlm-nih-gov.ezproxy.u-pec.fr/geo/query/acc.cgi?acc=GSE171169 - done
note: no controls, only 5d/14 days. Guo et al also published but data is unavailable
i) PS2APP and hMAPT.P301L - done (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE153895)
k) cuprizone 12 wk, 4 wk - done (GSE120744 https://www-ncbi-nlm-nih-gov.ezproxy.u-pec.fr/geo/query/acc.cgi?acc=GSE120744)
o) hSOD1.G93A (65/100 d) - done GSE178693 (https://www.sciencedirect.com/science/article/pii/S0969996120301522
m) repopulating/infiltrating microglia - done
  irradiation: GSE149247 https://www-ncbi-nlm-nih-gov.ezproxy.u-pec.fr/geo/query/acc.cgi?acc=GSE149247
  csf1r inhibition: GSE150169
  https://www-ncbi-nlm-nih-gov.ezproxy.u-pec.fr/geo/query/acc.cgi?acc=GSE150169


b) LCMV - none

g) hMapt.p301s

l)  MFP2ko

  also GSE133773
  https://www-ncbi-nlm-nih-gov.ezproxy.u-pec.fr/geo/query/acc.cgi?acc=GSE133773
  also GSE108416
  https://www-ncbi-nlm-nih-gov.ezproxy.u-pec.fr/geo/query/acc.cgi?acc=GSE108416
  other models: https://www-ncbi-nlm-nih-gov.ezproxy.u-pec.fr/geo/query/acc.cgi?acc=GSE103334 
https://www-ncbi-nlm-nih-gov.ezproxy.u-pec.fr/geo/query/acc.cgi?acc=GSE98969


```{r}
for(gene in allgenes){
  setEPS()
  postscript(paste0("~/Desktop/",gene,".eps"))
  x = VlnPlot(sobject,gene,group.by = "finalclusters")
  print(x)
  dev.off()
}

d = GetAssayData(sobject,slot = "data",assay = "RNA")[allgenes,]

s = data.frame(x = allgenes,row.names = allgenes)

for (group in levels(sobject$finalclusters)){
  cells = colnames(sobject)[sobject$finalclusters == group]
  sub = d[,colnames(d) %in% cells]
  s[paste0("median",group)] = apply(sub,1,median)
  s[paste0("mean",group)] = apply(sub,1,mean)
  sub[sub>0] = 1
  s[paste0("pct",group)] = apply(sub,1, function(x)sum(x)/length(cells))
}

new = c()
for (i in 1:nrow(lyso)){
  new = c(new,rep(lyso$Genotype[i],lyso$Count[i]))
}

write.csv(new,"~/Desktop/lysogenotype.csv")
```

