---
title: "Seurat_Part3_Clustering.Rmd"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

title: "Seurat part 3"
output: html_notebook
---
To get access to private functions: seurat:::Rts
genenames are the rows

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

(Part 5 from workshop)
```{r}
#source("https://bioconductor.org/biocLite.R")
#biocLite("BiocUpgrade")
library(Seurat)
dir<-"~/Desktop/LD_AVM02/Data"
setwd(dir)
library(ggplot2)
filename="MG_umap_normsep2.RDAta"
load(file=filename)
MG
```


Seurat implements an graph-based clustering approach. Distances between the cells are calculated based on previously identified PCs. Seurat approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNAseq data. Briefly, Seurat identify clusters of cells by a shared nearest neighbor (SNN) modularity optimization based clustering algorithm. First calculate k-nearest neighbors (KNN) and construct the SNN graph. Then optimize the modularity function to determine clusters. For a full description of the algorithms, see Waltman and van Eck (2013) The European Physical Journal B.

The FindClusters function implements the procedure, and contains a resolution parameter that sets the granularity of the downstream clustering, with increased values leading to a greater number of clusters. I tend to like to perform a series of resolutions, investigate and choose.

in parallel to t-sne; doesn't use tsne at all in the clustering algorithm

Resolution: telling the clustering algorithm how to group cells (similar to cuttree) - how far from the root do you cut the tree? low resolution means fewer clusters. 

Saves the clusters inside experiment.aggregate, can see them with "PrintFindClustersParams" (last line below)




Lets first investigate how many clusters each resolution produces and set it to the smallest resolutions of 0.5 (fewest clusters). finaly let's produce a table of cluster to sample assignments.
Note that the t-SNE plot never changes; you are only changing the mapped cluster colors. 

1 - find within the column names of metadata (i.e. types of meta data) anything that contains "res". Metadata now has columns with each resolution (0.5 - 4, 1 column each). In that column is stored the identity of the cluster to which the cells belong. 
Pick out each resolution column and find the length of unique identifiers in the metadata of that column. That produces a table with resolutions and number of clusters for each one. 
2- set the "identity" object to the cluster according to the resolution you chose (in this case 0.5). Default identity is the sample name. Once you run findclusters, it sets the identity to the cluster according to the lowest resolution. 



Now we will make a table that shows how many cells from each sample are in each cluster, allowing us to assign cluster identities ("Dep", "Ctrl", or "Both") to each cell based on the difference between the ratio of dep/ctrl and the expected ratio. 

This is totally optional, this was my way of assigning a name to a cluster that might not totally overlap 100% with the original identity of every cell in the cluster. 
```{r}
#Interesting clusters are microglia_11 (dep5), microglia_16 (P7/ctrl), microglia_9 (ctrl), neurons_4 (ctrl5), oligo10 (down in DP7) just in terms of number of cells - I could always do chi square tests just on those by pooling the rest
#microglia3 might be unhealthy
c = as.factor(MG$seurat_clusters)
levels(c)
all = 0:17
dep5 = c(11)
ctrl5 = c(4)
ctrl = c(9)
ctrl7 = c(16)
p5ctrl7 = c(10)


x = 1:18
x[all+1]="all"
x[dep5+1]="dep5"
x[ctrl5+1]="ctrl5"
x[ctrl+1]="ctrl"
x[ctrl7+1]="ctrl7"
x[p5ctrl7+1]="p5ctrl7"
levels(c) = x
MG$bias = c
table(MG$bias,MG$celltypecluster)
```
```{r}
table(MG$bias,MG$sample_description)
```
tSNE dimensionality reduction plots are then used to visualise clustering results. As input to the tSNE, you should use the same PCs as input to the clustering analysis.

default to the color of the t-SNE plot is to use whatever is set as the "identity"


Plot TSNE coloring by the slot 'ident' (default).

If you select dark.theme=T, you can change to a black background. 

```{r}
DimPlot(MG, reduction = "pca", dims = c(1,2), group.by = "sample_description")
DimPlot(MG, reduction = "pca", dims = c(1,2), group.by = "celltype")
DimPlot(MG, reduction = "pca", dims = c(1,2), group.by = "bias")
```

Plot TSNE coloring by the slot 'orig.ident' (sample names). using group.by variable to set coloring. Note that the actual (b&w) plot remains the same
```{r}
DimHeatmap(MG, dims = 1, cells = 500, balanced = T)
DimHeatmap(MG, dims = 1:15, cells = 500, balanced = T)
```


Plot TSNE coloring by the clustering resolution 4.0. Tsne plot is used for categorical data. do.label gives you the cluster ID in the legend. 


```{r}
DimPlot(MG,reduction = "umap",group.by="celltypecluster",label = TRUE)
DimPlot(MG,reduction = "umap",group.by="bias") 

```


FeaturePlot can be used to color cells with a 'feature', non categorical data, like number of UMIs. (colors with heat map type coloring - continuous)


Cluster 1 step beyond what you want, then merge the clusters that you think are the same clusters. That way you can group all the clusters within cell types you don't care as much about, and pull out individual clusters from the cell type of interest. 

Subsetting a cluster - if you pull out the cells from only one cluster, you could subset it differently by re-running most variable genes, PCA, t-SNE in case you want to subdivide. 
--Building a tree

```{r}
Idents(MG) = MG$celltypecluster
MG<- BuildClusterTree(
  MG,dims = 1:30,
  reorder = T,
  reorder.numeric = F)

PlotClusterTree(MG, cex=1, use.edge.length=F,node.depth=2)
library(ape)
```

Plot the split at node 30. Note that all the clusters are numbered 0-16 (for 17 clusters), nodes would then be numbered 18-30 (for example). This will only color the right, left sides of node 30 on the tsne


By setting new.ident.name, you can combine clusters that are already similar to each other. You can combine clusters to cluster together cells that you don't care about as much. 
Don't use mergenode - it is not working. 


Seurat can help you find markers that define clusters via differential expression.

FindMarkers identifies markers for a cluster relative to all other clusters.

FindAllMarkers does so for all clusters

FindAllMarkersNode defines all markers that split a Node (Warning: need to validate)

Specify the group you are interested in (in this case the combined 0,14 cluster) that distinguishes this group from everything else. You can also compare two groups using ident.2 = c(3) for example. 

Default - wilcoxon non-parametric t-test (haha not really a t-test) between two groups
FindMarkers uses the norm data, and some of the p-values are really high

pct.1 and pct.2 are the proportion of cells with expression above 0 in ident.1 and ident.2 respectively. p_val is the raw p_value associated with the differential expression test with adjusted value in p_val_adj. avg_logFC is the average log fold change difference between the two groups.

avg_diff (lines 130, 193 and) appears to be the difference in log(x = mean(x = exp(x = x) - 1) + 1) between groups. It doesn’t seem like this should work out to be the signed ratio of pct.1 to pct.2 so I must be missing something. It doesn’t seem to be related at all to how the p-values are calculated so maybe it doesn’t matter so much, and the sign is probably going to be pretty robust to how expression is measured.

These all seem to be expressed in different subsets of glia
Can use a violin plot to visualize the expression pattern of some markers
First 2 genes across the whole dataset 

Or a feature plot


FindAllMarkers can be used to automate the process across all clusters. WARNING: TAKES A LONG TIME TO RUN

```{r}
setwd("~/Desktop/plots")
markers_all <- FindAllMarkers(
    object = mgAVM02, 
    only.pos = FALSE, 
    min.pct = 0.10, #gene must be present in 10% of the cells in the cluster
    logfc.threshold = 0.15,
    test.use = "MAST"
)

dim(markers_all)
head(markers_all)
write.csv(markers_all,"allmarkersMG_0212.csv")
table(table(markers_all$gene)) #basically a histogram of how many genes are present in how many clusters
pcut = 1e-50
#setwd("~/Desktop/plots")
markers_all_single <- markers_all[markers_all$avg_logFC > 0,]
markers_all_single <- markers_all_single[markers_all_single$p_val_adj < pcut,]
markers_all_single <- markers_all_single[markers_all_single$gene %in% names(table(markers_all_single$gene))[table(markers_all_single$gene) == 1],] #markers that define a single cluster
write.csv(markers_all_single,"singlemarkersMG_0920.csv")
dim(markers_all_single)
table(table(markers_all_single$gene))

table(markers_all_single$cluster)

head(markers_all_single)

```
Plot a heatmap of genes by cluster for the top 5 marker genes per cluster

Make a metadata column with each celltype associated with all of its cells

Make a heatmap with the top 5 genes for every cluster

```{r}
#biocLite("dplyr")
library(dplyr)
top5 <- markers_all %>% group_by(cluster) %>% top_n(5, avg_logFC)
dim(top5)
DoHeatmap(
    object = mgAVM02, 
    features = top5$gene,size = 1
) + NoLegend()

setEPS()
postscript("~/Desktop/heatmap_mg_pval50.eps")
DoHeatmap(
    object = sobject, 
    features = top5$gene,
    group.by = 'seurat_clusters',
    size = 5,
    label = T,
    draw.lines = T
)

dev.off()
```
 

```{r}
DimPlot(object = MG,reduction = "umap", pt.size=1.0,label=T)
DimPlot(object = MG,reduction = "umap", pt.size=1.0,group.by = "fulldescription")
control<-names(MG$sample_description %in% "Control_P5")

#DimPlot(object=MG,cells = names(MG$fulldescription[MG$orig.ident=="LD5RC"]), pt.size=0.5,legend = "Control", group.by = "celltypecluster")
#DimPlot(object=MG,cells = names(MG$fulldescription[MG$orig.ident=="LD5LD"]), pt.size=0.5,legend = "Deprived", group.by = "celltypecluster")

DimPlot(object = MG,reduction = "umap", pt.size=0.5,split.by = "sample_description",group.by = "celltypecluster")
#cells.highlight will theoretically color cells by any category
```

Below is some code that you could use to find genes that are differentially expressed between only 2 clusters at a time. You can also use c() to find what differentiates one cluster from a set of other clusters. 


```{r}
#dep5 = 11,ctrl5 = 4, ctrl = 9, ctrl7 = 16,p5ctrl7 = 10, micro_3 might be poor quality
allmicro = c('microglia_0','microglia_1','microglia_2','microglia_3','microglia_5','microglia_7','microglia_8','microglia_13','microglia_17','microglia_9','microglia_16','microglia_11')

markersMicro11 = FindMarkers(MG, ident.1=c('microglia_11'), ident.2 = allmicro[1:11],only.pos=F)
dim(markersMicro11)
head(markersMicro11)
write.csv(markersMicro11,"markersdep5micro11.csv")

markersMicro4 = FindMarkers(MG, ident.1=c('neurons_4'),ident.2 = c('neurons_14','neurons_6'), only.pos=F)
dim(markersMicro4 )
head(markersMicro4 )
write.csv(markersMicro4 ,"markersctrl5neuro4.csv")

markersMicro9 = FindMarkers(MG, ident.1=c('microglia_9'),ident.2 = allmicro[c(1:9,11:12)], only.pos=F)
dim(markersMicro9 )
head(markersMicro9 )
write.csv(markersMicro9 ,"markersctrlmicro9.csv")

markersMicro16 = FindMarkers(MG, ident.1=c('microglia_16'),ident.2 = allmicro[c(1:10,12)], only.pos=F)
dim(markersMicro16)
head(markersMicro16)
write.csv(markersMicro16 ,"markersctrl7micro16.csv")

markersMicro10 = FindMarkers(MG, ident.1=c('oligo_10'), ident.2 = 'oligo_15',only.pos=F)
dim(markersMicro10)
head(markersMicro10)
write.csv(markersMicro10 ,"markersp5ctrl7oligo10.csv")
```

Extract the data, use limmaVoom to do differential expression analysis (some statisticians don't like deseq2 or the negative binomial distribution). (Said by the makers of limmaVoom). 

At this point in time you should use the tree, markers, domain knowledge, and goals to finalize your clusters. This may mean adjusting PCA to use, mergers clusters together, choosing a new resolutions, etc. When finished you can further name it cluster by something more informative. Ex.

Get expression of marker genes - means of each cluster by gene; what is the gene expression value for that cluster for each gene - note that we are using the normalized, data column, so there will be normalized values since the mean is set to zero

Save

```{r}
filename<-"MG_Part3.RData"
save(MG,file=filename)
```



```{r}
sessionInfo()
```




