---
title: "Seurat_Part3_Clustering.Rmd"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

title: "Seurat part 3"
output: html_notebook
---
To get access to private functions: seurat:::Rts
genenames are the rows

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

(Part 5 from workshop)
```{r}
source("https://bioconductor.org/biocLite.R")
#biocLite("BiocUpgrade")
library(Seurat)
library(ggplot2)
filename="~/Desktop/Sequencing/LD_RNASeq2018/RData_v3/LD1_sctransform1_v3.RDAta"
load(file=filename)
LD1
```


Seurat implements an graph-based clustering approach. Distances between the cells are calculated based on previously identified PCs. Seurat approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNAseq data. Briefly, Seurat identify clusters of cells by a shared nearest neighbor (SNN) modularity optimization based clustering algorithm. First calculate k-nearest neighbors (KNN) and construct the SNN graph. Then optimize the modularity function to determine clusters. For a full description of the algorithms, see Waltman and van Eck (2013) The European Physical Journal B.

The FindClusters function implements the procedure, and contains a resolution parameter that sets the granularity of the downstream clustering, with increased values leading to a greater number of clusters. I tend to like to perform a series of resolutions, investigate and choose.

in parallel to t-sne; doesn't use tsne at all in the clustering algorithm

Resolution: telling the clustering algorithm how to group cells (similar to cuttree) - how far from the root do you cut the tree? low resolution means fewer clusters. 

Saves the clusters inside LD1, can see them with "PrintFindClustersParams" (last line below)




Lets first investigate how many clusters each resolution produces and set it to the smallest resolutions of 0.5 (fewest clusters). finaly let's produce a table of cluster to sample assignments.
Note that the t-SNE plot never changes; you are only changing the mapped cluster colors. 

1 - find within the column names of metadata (i.e. types of meta data) anything that contains "res". Metadata now has columns with each resolution (0.5 - 4, 1 column each). In that column is stored the identity of the cluster to which the cells belong. 
Pick out each resolution column and find the length of unique identifiers in the metadata of that column. That produces a table with resolutions and number of clusters for each one. 
2- set the "identity" object to the cluster according to the resolution you chose (in this case 0.5). Default identity is the sample name. Once you run findclusters, it sets the identity to the cluster according to the lowest resolution. 



Now we will make a table that shows how many cells from each sample are in each cluster, allowing us to assign cluster identities ("Dep", "Ctrl", or "Both") to each cell based on the difference between the ratio of dep/ctrl and the expected ratio. 

This is totally optional, this was my way of assigning a name to a cluster that might not totally overlap 100% with the original identity of every cell in the cluster. 
```{r}
celltypecluster = paste0(LD1$celltype,"_",LD1$seurat_clusters)
LD1$celltypecluster = celltypecluster
table<-table(LD1$celltypecluster,LD1$sample_description)
depvctrl<-table[,2]/table[,1]
table<-cbind(table,depvctrl)
table
total<-table(LD1$sample_description)
ratio<-total[2]/total[1]
names(ratio)<-"expected"
ratio<-rep(ratio,length(depvctrl))
clusterbias<-as.data.frame(cbind(table,ratio))
```

```{r}
##For res = 1.5
celltypecluster = paste0(LD1$celltype1.5,"_",LD1$SCT_snn_res.1.5)
LD1$celltypecluster1.5 = celltypecluster
table<-table(LD1$celltypecluster1.5,LD1$sample_description)
depvctrl<-table[,2]/table[,1]
table<-cbind(table,depvctrl)
table
total<-table(LD1$sample_description)
ratio<-total[2]/total[1]
names(ratio)<-"expected"
ratio<-rep(ratio,length(depvctrl))
clusterbias<-as.data.frame(cbind(table,ratio))
clusterbias
```
```{r}
threshold=1.5 #define this how you like. This is going to be the minimum ratio between the expected random percentage of treated cells in a cluster and the actual percentage in order to define that cluster as control, deprived, or mixed. 
bias<-function(table){
  bias<-rep("no",nrow(table))
  for (i in 1:nrow(table)){
    bias[i]="Both"
    if (table[i,3]>threshold*table[i,4]){
      bias[i]="Dep"
    }
    else if (table[i,3]<table[i,4]/threshold){
      bias[i]="Ctrl"
    }
  }
  table<-cbind(table,bias)
}

#assign cluster identities based on ratio
clusterbias<-bias(clusterbias)
clusterbias
clusterbias[,c(3,5)]
clusterid<-as.factor(LD1$celltypecluster1.5)
levels(clusterid)<-clusterbias$bias
cellbias<-clusterid
LD1$bias1.5 = cellbias
table(LD1$bias1.5,LD1$sample_description)
clusterbias = paste0(LD1$celltypecluster,"_",LD1$bias1.5)
LD1$clusterbias1.5 = clusterbias
```

tSNE dimensionality reduction plots are then used to visualise clustering results. As input to the tSNE, you should use the same PCs as input to the clustering analysis.

default to the color of the t-SNE plot is to use whatever is set as the "identity"


Plot TSNE coloring by the slot 'ident' (default).

If you select dark.theme=T, you can change to a black background. 

Plot TSNE coloring by the slot 'orig.ident' (sample names). using group.by variable to set coloring. Note that the actual (b&w) plot remains the same
```{r}
DimHeatmap(LD1, dims = 1, cells = 500, balanced = T)
DimHeatmap(LD1, dims = 1:15, cells = 500, balanced = T)
```

```{r}
getwd()
Idents(LD1) = "celltypecluster"
LD1<- BuildClusterTree(LD1,dims = 1:30)

PlotClusterTree(LD1, cex=1, use.edge.length=F,node.depth=2)
library(ape)


tree1.5 = LD1@tools$BuildClusterTree
setEPS()
postscript("tree_LD1.5.eps")
plot.phylo(tree1.5, use.edge.length = T, direction = "rightwards")
dev.off()

o = c("neurons 14","pericytes 13","oligos 7","endothelial 4", "microglia 8","microglia 11", "microglia 2","microglia 6","microglia 3", "microglia 9", "astrocytes 0","astrocytes 5","neurons 12","opcs 1","opcs 10")
```

FindAllMarkers can be used to automate the process across all clusters. WARNING: TAKES A LONG TIME TO RUN

```{r}
markers_all <- FindAllMarkers(
    object = LD1, 
    only.pos = TRUE, 
    min.pct = 0.25, #gene must be present in 25% of the cells in the cluster
    logfc.threshold = 0.25
)

dim(markers_all)
head(markers_all)
write.csv(markers_all,"RNUMIallmarkerssortonllowfilternobatch.csv")
table(table(markers_all$gene)) #basically a histogram of how many genes are present in how many clusters

markers_all_single <- markers_all[markers_all$gene %in% names(table(markers_all$gene))[table(markers_all$gene) == 1],] #markers that define a single cluster
write.csv(markers_all_single,"RNUMIsinglemarkerssortonllowfilternobatch.csv")
dim(markers_all_single)
table(table(markers_all_single$gene))

table(markers_all_single$cluster)

head(markers_all_single)
```
Plot a heatmap of genes by cluster for the top 5 marker genes per cluster

Make a metadata column with each celltype associated with all of its cells

Make a heatmap with the top 5 genes for every cluster

```{r}
#biocLite("dplyr")
library(dplyr)
top5 <- markers_all %>% group_by(cluster) %>% top_n(5, avg_logFC)
dim(top5)
DoHeatmap(
    object = LD1, 
    features = top5$gene,
) + NoLegend()
```
 

```{r}
r = table(LD1$celltypecluster1.5,LD1$sample_description)
Control = r[,1]*100/colSums(r)[1] #calculate what percent of sample 1 each cluster makes up
Deprived = r[,2]*100/colSums(r)[2]

relconpct = cbind(Control,Deprived)
relconpct

x = rev(o)
relconordered = relconpct[order(relconpct[,1],decreasing = T),]
relconpct = relconpct[match(o,rownames(relconpct)),]
setEPS()
postscript("clustercomppctall.eps")
barplot(relconpct, main="Cluster composition by percent of sample",
        xlab="Cluster", ylab = "Number of Cells", ylim = c(0,100), col=c("darkblue","lightblue","red","orange","purple","lightgrey","darkgrey","magenta","yellow","black","red3","cornflowerblue","grey"),axisnames = T,
        width = .2,xlim = c(0,2),legend = rownames(relconpct), space = 0.6,cex.names = 0.7,axis.lty = 1)
dev.off()
getwd()
#cells.highlight will theoretically color cells by any category
```
```{r}
#split the plot by celltype
r = table(LD1$celltypecluster,LD1$sample_description)
#normalize each sample to the same total number of cells (1000)
Control = (r[,1]/colSums(r)[1])*1000
Deprived = (r[,2]/colSums(r)[2])*1000
r = cbind(Control,Deprived)
#r = rbind(r,total = colSums(r))
ra = r[c(1:2),]
ra = cbind(ra,total = rowSums(ra))
rm = r[c(4:7),]
rm = cbind(rm,total = rowSums(rm))
re = r[c(3),]
re = c(re,total = sum(re))
ro = r[c(10:12),]
ro = cbind(ro,total = rowSums(ro))

ca = c("yellow","darkgoldenrod1")
cm = c("deeppink","indianred1","red","firebrick") #deeppink
ce = c("darkturquoise")
co = c("cornflowerblue","darkolivegreen3","forestgreen")

r = rm
cols = c("darkblue","red3")
for (row in 1:nrow(r)){
  r[row,] = r[row,]/r[row,ncol(r)]*100
  r
}
#Deprived = r[1:(nrow(r)-1),2]*100/r[nrow(r),2]

#Control = r[1:(nrow(r)-1),1]*100/sum(r[nrow(r),1:2])
#Deprived = r[1:(nrow(r)-1),2]*100/sum(r[nrow(r),1:2])
relconpct = r[,1:2]
relconpct = t(relconpct)
#o = c("neurons_12_Dep","pericytes_11_Dep","oligos_7_Both","endothelial_4_Both", "microglia_8_Dep","microglia_5_Both", "microglia_0_Dep","microglia_3_Ctrl","astrocytes_1_Dep","astrocytes_6_Ctrl","neurons_10_Both","opcs_2_Both","opcs_9_Dep")
#relconordered = relconpct[order(relconpct[,1],decreasing = T),]
#relconpct = relconpct[match(o,rownames(relconpct)),]
setEPS()
postscript("clustercomppctmicro_2.eps")
barplot(relconpct, main="Cluster composition by percent of celltype",
        xlab="Cluster", ylab = "% of cluster", ylim = c(0,100), col=cols,axisnames = T,
        width = .2,xlim = c(0,2),legend = rownames(relconpct), space = 0.6,cex.names = 0.7,axis.lty = 1)
dev.off()
getwd()
#cells.highlight will theoretically color cells by any category
```


```{r}
#Try flipping the plot
r = table(LD1$sample_description,LD1$clusterbias)

Deprived = round(r[2,]/1.69)
r[2,] = Deprived
total = colSums(r)
Control = r[1,]*100/colSums(r) #calculate what percent of each cluster is occupied by a given sample
Deprived = r[2,]*100/colSums(r)

relconpct = rbind(Control,Deprived)
relconpct
o = c("neurons_12_Dep","pericytes_11_Dep","oligos_7_Both","endothelial_4_Both", "microglia_8_Dep","microglia_5_Both", "microglia_0_Dep","microglia_3_Ctrl","astrocytes_1_Dep","astrocytes_6_Ctrl","neurons_10_Both","opcs_2_Both","opcs_9_Dep")
relconpct = relconpct[,match(o,colnames(relconpct))]
setEPS()
postscript("clustercomppctallhoriz.eps")

barplot(relconpct, main="Cluster composition by percent of sample",
        xlab="Percent of cluster", ylab = "Cluster", col=c("darkblue","red3"),axisnames = T,
        width = .2,legend = rownames(relconpct),horiz = T,xlim = c(0,150))
dev.off()
getwd()
#cells.highlight will theoretically color cells by any category
```
Below is some code that you could use to find genes that are differentially expressed between only 2 clusters at a time. You can also use c() to find what differentiates one cluster from a set of other clusters. 
```{r}
Idents(LD1) = "clusterbias"
markersAstrodep= FindMarkers(LD1, ident.1="astrocytes_1_Dep", ident.2="astrocytes_6_Ctrl",test.use ="MAST",only.pos=F,logfc.threshold = 0,min.pct = 0.01)
dim(markersAstrodep)
head(markersAstrodep)
write.csv(markersAstrodep,"AstroDepvsCtrlnew.csv")

markersOPCdepvboth = FindMarkers(LD1, ident.1=c('opcs_9_Dep'), ident.2='opcs_2_Both',test.use ="MAST",only.pos=F,logfc.threshold = 0,min.pct = 0.01)
dim(markersOPCdepvboth)
head(markersOPCdepvboth)
write.csv(markersOPCdepvboth,"opcdepvboth.csv")

subid = paste0(LD1$sample_description,"_in_",LD1$celltypecluster)
LD1$subid = subid
Idents(LD1) = subid
markersOPCdepvctrl9 = FindMarkers(LD1, ident.1=c('Deprived_in_opcs_9'), ident.2='Control_in_opcs_9',test.use ="MAST",only.pos=F,logfc.threshold = 0,min.pct = 0.01)
dim(markersOPCdepvctrl9)
head(markersOPCdepvctrl9)
write.csv(markersOPCdepvctrl9,"opcdepvctrl9.csv")

markersOPCdepvctrl2 = FindMarkers(LD1, ident.1=c('Deprived_in_opcs_2'), ident.2='Control_in_opcs_2',test.use ="MAST",only.pos=F,logfc.threshold = 0,min.pct = 0.01)
dim(markersOPCdepvctrl2)
head(markersOPCdepvctrl2)
write.csv(markersOPCdepvctrl2,"opcdepvctrl2.csv")
```

Extract the data, use limmaVoom to do differential expression analysis (some statisticians don't like deseq2 or the negative binomial distribution). (Said by the makers of limmaVoom). 

At this point in time you should use the tree, markers, domain knowledge, and goals to finalize your clusters. This may mean adjusting PCA to use, mergers clusters together, choosing a new resolutions, etc. When finished you can further name it cluster by something more informative. Ex.



```{r}
setEPS()
postscript("umap_celltype.eps")
DimPlot(LD1,label=T,reduction = "umap",group.by = "celltype")
dev.off()

setEPS()
postscript("umap_cellbias1.5.eps")
DimPlot(LD1,reduction = "umap",group.by = "celltypecluster1.5")
dev.off()

setEPS()
postscript("vln_Cx3cr1.eps")
VlnPlot(LD1,c("Cx3cr1"),group.by = "celltype",sort = "increasing")
dev.off()

setEPS()
postscript("vln_Aldh1l1.eps")
VlnPlot(LD1,c("Aldh1l1"),group.by = "celltype",sort = "increasing")
dev.off()

setEPS()
postscript("vln_Pdgfra.eps")
VlnPlot(LD1,c("Pdgfra"),group.by = "celltype",sort = "increasing")
dev.off()

setEPS()
postscript("vln_Tuj1.eps")
VlnPlot(LD1,c("Tubb3"),group.by = "celltype",sort = "increasing")
dev.off()


setEPS()
postscript("vln_Mbp.eps")
VlnPlot(LD1,c("Mbp"),group.by = "celltype",sort = "increasing")
dev.off()


setEPS()
postscript("vln_Cd31.eps")
VlnPlot(LD1,c("Pecam1"),group.by = "celltype",sort = "increasing")
dev.off()


setEPS()
postscript("vln_Ng2.eps")
VlnPlot(LD1,c("Cspg4"),group.by = "celltype",sort = "increasing")
dev.off()

setEPS()
postscript("QC_umap_sex.eps")
DimPlot(LD1,reduction = "umap",group.by = "sex")
dev.off()
```


 Get expression of marker genes - means of each cluster by gene; what is the gene expression value for that cluster for each gene - note that we are using the normalized, data column, so there will be normalized values since the mean is set to zero


```{r}
setEPS()
postscript("Il4ra_Il13ra1_Jerika.eps")
VlnPlot(LD1,features = c("Il4ra","Il13ra1"),group.by = "celltype",slot = "data")
dev.off()
```

```{r}
#VolcanoPlot

fc = read.csv("~/Desktop/Sequencing/LD_RNASeq2018/Spreadsheets/opcdepvctrl9.csv",stringsAsFactors = F)
colnames(fc)[1] = "Gene"
fc = fc[!is.na(fc$avg_logFC),]
colorkeysdown = fc$Gene[fc$avg_logFC < -log2(1.3) & fc$p_val_adj < 10e-3]
colorkeysup = fc$Gene[fc$avg_logFC > log2(1.3) & fc$p_val_adj < 10e-3]
allcolors = rep("darkgrey",length(fc$Gene))
names(allcolors) = fc$Gene
allcolors[names(allcolors) %in% colorkeysdown] = "blue"
allcolors[names(allcolors) %in% colorkeysup]= "red"
names(allcolors)[allcolors == "blue"] = "d"
names(allcolors)[allcolors == "red"] = "u"
names(allcolors)[allcolors == "darkgrey"] = "-"

library(EnhancedVolcano)
EnhancedVolcano(fc,
lab = fc$Gene,
x = 'avg_logFC',
y = 'p_val_adj',
xlim = c(-1, 1),
title = 'OPC DepvCtrl Cluster9',
subtitle = "",
drawConnectors = F,
legendPosition = 'right',
legendVisible = F,
pCutoff = 10e-3,
FCcutoff = log2(1.3),
transcriptPointSize = 1.5,
transcriptLabSize = 3.0,
col=c('black', 'black', 'black', 'red3'),
colCustom = allcolors,
gridlines.major = F,
gridlines.minor = F,
colAlpha = 1)


setEPS()
postscript("OPC DepvCtrl Cluster9.eps")
EnhancedVolcano(fc,
lab = fc$Gene,
x = 'avg_logFC',
y = 'p_val_adj',
xlim = c(-1, 1),
title = 'OPC DepvCtrl Cluster9',
subtitle = "",
drawConnectors = F,
legendPosition = 'right',
legendVisible = F,
pCutoff = 10e-3,
FCcutoff = log2(1.3),
transcriptPointSize = 1.5,
transcriptLabSize = 3.0,
col=c('black', 'black', 'black', 'red3'),
colCustom = allcolors,
gridlines.major = F,
gridlines.minor = F,
colAlpha = 1)

dev.off()
```



Save

```{r}
filename<-"LD1_part3.RData"
save(LD1,file=filename)
```



```{r}
sessionInfo()
```




